lib/addonutils.cpp

```diff
@@ -0,0 +1,189 @@
+/*
+ * Cppcheck - A tool for static C/C++ code analysis
+ * Copyright (C) 2007-2019 Cppcheck team.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "addonutils.h"
+
+#include "path.h"
+#include "utils.h"
+
+#define PICOJSON_USE_INT64
+#include <picojson.h>
+#include <fstream>
+#include <memory>
+#include <string>
+#include <sstream>
+
+static inline void throwAddonError(const std::string &msg)
+{
+    throw InternalError(nullptr, msg, InternalError::ADDON_ERROR);
+}
+
+Addon::Addon(const std::string &addonFileName, const std::string &exename)
+    : mArgs("")
+{
+    getAddonInfo(addonFileName, exename);
+}
+
+Addon::~Addon() {}
+
+void Addon::getAddonInfo(const std::string &fileName, const std::string &exename)
+{
+    if (fileName.find(".") == std::string::npos)
+        return getAddonInfo(fileName + ".py", exename);
+
+    if (endsWith(fileName, ".py", 3)) {
+        mScriptFile = getFullPath(fileName, exename);
+        if (mScriptFile.empty())
+            throwAddonError("Did not find addon " + fileName);
+
+        std::string::size_type pos1 = mScriptFile.rfind("/");
+        if (pos1 == std::string::npos)
+            pos1 = 0;
+        else
+            pos1++;
+        std::string::size_type pos2 = mScriptFile.rfind(".");
+        if (pos2 < pos1)
+            pos2 = std::string::npos;
+        mName = mScriptFile.substr(pos1, pos2 - pos1);
+
+        return;
+    } else {
+        if (!endsWith(fileName, ".json", 5))
+            throwAddonError("Failed to open addon " + fileName);
+
+        std::ifstream fin(fileName);
+        if (!fin.is_open())
+            throwAddonError("Failed to open " + fileName);
+
+        picojson::value json;
+        fin >> json;
```

Comment:
Do not forget to merge the fix adding json-parser error checking here ;)

Reply:
Thanks, I'll take a look at this. Really I'm waiting for next release, to merge master changes in my branch. And somehow fix windows build.